### ISSUE TITLE
Proposal: Disable CS1591 "Missing XML comment" on generated code

### ISSUE THREAD

**Issue by SolalPirelli** [dotnet/roslyn#12702]
Reporting CS1591 on generated code (i.e. classes marked with `GeneratedCodeAttribute`) is annoying, since the code may be re-generated at each build, and rather pointless anyway since generated code is rarely useful to document.

For instance, UWP XAML generates a bunch of stuff in a "XamlTypeInfo.g.cs" file, including a public type which does not have XML doc; this means that by default an UWP library project (for which XML doc is useful) will not build with warning as errors.

**Comment by miloush** [comment]
I would prefer the XML documentation to be generated instead.

**Comment by SolalPirelli** [comment]
> I would prefer the XML documentation to be generated instead.

So would I, but changing every existing code generator is not feasible.

**Comment by BrainSlugs83** [comment]
This happens with settings files too (and a bunch of other places), it's extremely annoying, as we have no power to comment those files, and I really don't want to turn off CS1591. -- In that case, the only thing that makes sense is to disable the warning for those files (because how would the generator know what comment to put for my custom settings?).

**Comment by BrainSlugs83** [comment]
On a side note, it also makes "Treat Warnings as Errors" completely useless too.  (i.e. if you use out of the box settings, turn on treat warnings as errors, and XML documentation -- it will be impossible to compile your project without manually editing the generated code files).

**Comment by DRAirey1** [comment]
Amen.  This is a big issue for Windows 8/10 development, especially with the generated code for type information.

**Comment by jaredpar** [comment]
Unfortunately the notion of generated code is generally a heuristic instead of a solid item.  There are too many attributes / conventions to account for.  That makes it harder for us to change compiler behavior based on it. 

Additionally there is a very simple work around for generators who explicitly don't want this warning.  Simply generate the following into the code file:

``` c#
#pragma warning disable 1591
```

**Comment by SolalPirelli** [comment]
If I were the one generating the code, sure, but in the case of UWP XAML, it's definitely not me... Otherwise I might as well generate the XML doc.

**Comment by DRAirey1** [comment]
I strongly disagree.  You will catch 95% of the issues by checking for the text "auto-generated" at the start of the module.

**Comment by DRAirey1** [comment]
>> Additionally there is a very simple work around for generators who explicitly don't want this warning. Simply generate the following into the code file:
>> #pragma warning disable 1591
Have you ever worked with Windows 8 or Windows 10?  There's a significant chunk of code that's generated in order to get around the limitations of the RTL and that all gets regurgitated back at you when you build.

**Comment by Portikus** [comment]
So what? This wont be fixed? We have to deal with warnings for every UWP *.g.cs file which has public non commented members? The huge amount of warnings is killing me and make it quite impossible to see real warnings. 
Also I'm using SimpleStubs which generates a huge amount of stubs in my project. All not commented..
@DRAirey1 is right :

> You will catch 95% of the issues by checking for the text "auto-generated" at the start of the module

**Comment by sharwell** [comment]
@SolalPirelli @BrainSlugs83 @DRAirey1 @Portikus I've dealt with a great deal of different situations related to this as part of my work on StyleCop Analyzers. While I don't have a "one size fits all" answer, if you get stuck on a *particular* case please feel free to call my attention to it and I can help you work through it so it's no longer a problem.

**Comment by SolalPirelli** [comment]
> What kind of project are you working in?

UWP libraries

> What specific warnings are you seeing that you would like to avoid?

CS1591 "Missing XML comment"

> What is the source of the file where the warning appears 

XamlTypeInfo.g.cs generated by UWP tools for XAML files.

**Comment by DRAirey1** [comment]
Sam, would you please explain to us disabling 1591 whenever you see <auto-generated> in the header of a file won't solve 95% of the issues here.

**Comment by sharwell** [comment]
@DRAirey1 I don't have any more say on the rules by which a compiler warning are reported than you have. What I do have is experience working around things that might otherwise block people. 😄

**Comment by Portikus** [comment]
>What kind of project are you working in?
UWP Libraries and Apps
>What specific warnings are you seeing that you would like to avoid?
CS1591 "Missing XML comment"
>What is the source of the file where the warning appears 
*.g.cs like MainPage.g.cs e.g for the Method "GetBindingConnector". 
Reproduce: 
Just create a Blank App in Visual Studio 2017 and check the checkbox XML documentation file.

**Comment by Portikus** [comment]
>What kind of project are you working in?

.NetStandard
>What specific warnings are you seeing that you would like to avoid?

CS1591 "Missing XML comment"

>What is the source of the file where the warning appears 

SimpleStubs.generated.cs from SimpleStubs version 2.3.3.
https://www.nuget.org/packages/Etg.SimpleStubs/
https://github.com/Microsoft/SimpleStubs

**Comment by sharwell** [comment]
> SimpleStubs.generated.cs from SimpleStubs version 2.3.3.

@Portikus I would recommend you disable CS1591 and CS1573 in test projects using one of the strategies documented in the [How to fix violations](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA0001.md#how-to-fix-violations) section of StyleCop Analyzers rule SA0001. If you cannot do that for some reason, then the solution here would be:

1. Request SimpleStubs add a `#pragma warning suppress 1591` to the generated code
2. Use a GlobalSuppressions.cs file in your test project to selectively disable CS1591 on that generated code

**Comment by DRAirey1** [comment]
Sam, you're late to the conversation.  We all know you can do this manually.  Digging into generated code in the object directories and updating what is supposed to be invisible to programmers is a time-consuming task that is prone to errors.  The end result is we don't want to see warnings that generated code hasn't been documented.  I don't think you or SolalPirelli realizes how often this occurs when building Modern UI applications.

**Comment by sharwell** [comment]
@SolalPirelli @Portikus  I filed an internal bug for the XAML compiler warnings:

![image](https://user-images.githubusercontent.com/1408396/27188931-631fd526-51b5-11e7-969b-7fe80d8a9681.png)

**Comment by Portikus** [comment]
@sharwell thank you, the global suppression file will help me with SimpleStubs. 
Will we get noticed about updates to the topic?

**Comment by DRAirey1** [comment]
Actually, I just got through looking at Sam's suggestion.  Turns out I created a similar solution and posted it as a package.  You can check it out at:

https://www.nuget.org/packages/IgnoreGeneratedFilesBuildTask/

**Comment by sharwell** [comment]
@DRAirey1 I filed dotnet/roslyn#20242. It's a higher-level feature than the compiler that has come up in multiple discussions. Among other things, it would allow you to create an analyzer package that included logic to suppress warnings (from the compiler or from an analyzer) using logic of your choosing.

**Comment by sharwell** [comment]
@Portikus 

> Will we get noticed about updates to the topic?

Probably not automatically, but if you aren't seeing progress you can always ask using the Bug ID from that image.

**Comment by Portikus** [comment]
@sharwell  

> 2.   Use a GlobalSuppressions.cs file in your test project to selectively disable CS1591 on that generated code

I think the suggestion with the Supression file wont work me. SimpleStubs generating many classes in many namespaces in one file. I am not able to supress all CS1591 warnings by my own.

**Comment by rynowak** [comment]
To add my feedback to this issue, as an owner of a code generator (ASP.NET Core Razor) this doesn't scale for us when warnings like `CS1591` are part of the project's configuration. 

We can't test for (and react to) every way that users can have their project configured. If more of these kinds of warnings are going to be added in the future (based on project settings, not related to the correctness of code) then it just makes more work for tool authors. 

It would be great to have `#pragma generated` or something like that that's clearer about intent. [Yeah yeah, I know](https://xkcd.com/927/).

**Comment by SolalPirelli** [comment]
> It would be great to have #pragma generated or something like that that's clearer about intent.

Why not `GeneratedCodeAttribute`? It already exists. (although I have no idea if it's in .NET Standard)

**Comment by sharwell** [comment]
@SolalPirelli The attribute can't be used to distinguish very well between generated code in a `partial class` and user code for the same class in a different file. For StyleCop Analyzers 1.0 (which handled its own generated code detection) we removed all special cases related to this attribute for a big performance win. I'm concerned that in 1.1 we may have introduced a performance regression by allowing Roslyn (which checks this attribute) to check this for us, so I may end up needing to remove the special cases for it there too.

**Comment by rynowak** [comment]
> Why not GeneratedCodeAttribute? It already exists. (although I have no idea if it's in .NET Standard)

I don't actually feel super invested in what exactly we have to do to codegen-wise to accomplish this. The Roslyn folks are in the best position to come up with a solution.

The docs for `#pragma` say:
>#pragma gives the compiler special instructions for the compilation of the **file in which it appears.**

This seems the cleanest to me because it's not entangled with any other language features and is file-centric. But this is just me doing backseat language design. I'd be satisfied with anything other than playing whackamole with future warnings.

**Comment by BrainSlugs83** [comment]
I highly recommend supporting the `GeneratedCodeAttribute`, because it's what most libraries and code generation tools use today. -- It does us no good to tell us to add a `#pragma warning disable 1591` or any other `#pragma` directive for that matter, since we're obviously powerless to edit the generated code. :(

Adding a new pragma and telling us to use that, in the BEST case, means, that if you implement it today, our pain will go away in about 5 years. -- At the worst case, it's just another way to mark code as generated, and just another thing that the tools will have to check for. -- It's seriously an annoying proposition.

To my knowledge, you can't suppress CS1591 in the global suppressions file -- the VS 2015 tooling definitely won't give you the option anyway; it will only suppress with a `#pragma` directive for CS warnings. -- And obviously we don't want to turn the warning off for a whole project -- Just because MS isn't documenting their code anymore, doesn't mean we don't want to document ours! :(

**Comment by BrainSlugs83** [comment]
Alternately, giving us a way to suppress warnings in the .csproj for a specific file (or group of files) would be great. -- And while you're at it, hiding warnings from files that aren't even in the .csproj (i.e. *.g.cs) would be a step in the right direction...

**Comment by DRAirey1** [comment]
Seriously, the people that make these decisions do not develop for UWP.  Otherwise they would have realized the problem long ago.  Try my fix and let me know how it works out for you.

https://www.nuget.org/packages/IgnoreGeneratedFilesBuildTask/

I'd prefer it if they caught the problem at the root, but your complaints are falling on deaf ears.

**Comment by sharwell** [comment]
> To my knowledge, you can't suppress CS1591 in the global suppressions file

This is correct. The global suppressions file can be used to suppress warnings reported by analyzers, but not ones reported by the compiler. It bit me too while investigating this issue. I couldn't find an issue for this, so if it's something you'd like to see changed I would encourage you to file a separate issue specifically for it (at minimum you'd have an answer about why it currently behaves that way).

> And obviously we don't want to turn the warning off for a whole project

One interesting possibility does involve turning off this warning. StyleCop Analyzers has several rules related to documentation, several of which closely resemble behavior normally seen from CS1591. There are two big differences you would see from switching from relying on CS1591 to relying on a separate analyzer:

1. The documentation requirements are more configurable. CS1591 applies only to exposed elements, but StyleCop Analyzers has [several additional places](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/Configuration.md#documentation-requirements) where documentation can be marked as required.
2. Since StyleCop Analyzers is a set of Roslyn analyzers, it will not report warnings in generated code and supports global suppressions.

:memo: Note that I'm not trying to say this is the long-term solution you want. I'm only pointing it out because it is a possible solution which happens to be immediately available.

**Comment by sharwell** [comment]
> Seriously, the people that make these decisions do not develop for UWP.

This is a fascinating and common problem which is definitely not unique to this team or even to this company. Back when I worked in the game industry, it always amazed me *how much* better the players were at the games than the creators of the game who "know everything about it".

Every day I work with several tools and workflows to solve problems, yet I know that I'm only exercising a minute, almost insignificant, fraction of the use cases these tools are built to handle. I love reading about other workflows and the successes and challenges users have in them. Sometimes the reports involve something simple that we can just fix, and other times the reports require a lot of effort to understand a workflow that substantially deviates from our "comfort zone". Even though they can take a great deal of discussion before we figure out what to do, the latter are actually some of my favorite issues because they challenge the way I view daily developer activities and the ideas we put forward in attempts to improve everyone's productivity.

**tl;dr:** I don't know what the final fix to the issues presented above will be, but I am definitely interested in the problem space and want to see the pain points resolved.

**Comment by sharwell** [comment]
@SolalPirelli @Portikus My internal bug was moved to [Microsoft 13310683](https://microsoft.visualstudio.com/DefaultCollection/OS/_workitems/edit/13310683)), where it was closed as not reproducible. According to [Daniel Zilcsak](https://www.linkedin.com/in/danielzilcsak/):

> I tried this repro on SDK 16240 (a recent RS3 one) and I could not reproduce any errors in Mainpage.g.cs. The error I saw was in Mainpage.xaml.cs around the MainPage constructor, and as soon as I added a documentation comment, the solution built fine.
>
> Which SDK were you using? I remember we fixed a bunch of these in RS3.

This SDK may be newer than the one you have available, but either way you can consider this "fixed pending release", though I have absolutely no insight into exactly when that release would be.

**Comment by Portikus** [comment]
@sharwell thank you very much for the update. It seems with the newest Visual Studio version 15.3.3 that some of the generated code got summaries  :). But unfortunately some don't. 
There is a generated class:
`public sealed class XamlMetaDataProvider: global::Windows.UI.Xaml.Markup.IXamlMetadataProvider `
which has no code summary and therefore raises a CS1591 warning. Maybe you can reply that to Daniel Zilcsak. But I didn't figure out how to get that class generated.

**Comment by sharwell** [comment]
@Portikus thanks for the update. I left a comment on the internal bug. If you do find a clear way to reproduce the issue please let me know and I'll add that too. :smile:

**Comment by HugoRoss** [comment]
> > To my knowledge, you can't suppress CS1591 in the global suppressions file
> 
> This is correct. The global suppressions file can be used to suppress warnings reported by analyzers, but not ones reported by the compiler. It bit me too while investigating this issue. I couldn't find an issue for this, so if it's something you'd like to see changed I would encourage you to file a separate issue specifically for it (at minimum you'd have an answer about why it currently behaves that way).
> 
> > And obviously we don't want to turn the warning off for a whole project
> 

Question to the world: Is it possible to do the opposite as we try to do? Disable CS warnings like CS0618 and CS1591 for the whole project and have an analyzer to give me the warnings instead of the compiler, which would allow me to selectively exclude classes or namespaces of generated files?

**Comment by daaaasdev** [comment]
Hi @HugoRoss, Did you manage to do what you were describing

**Comment by HugoRoss** [comment]
I haven't tried yet. With Corona lock-down and kids home there is no time
left to do anything at all...

Am Mi., 1. Juli 2020 um 15:50 Uhr schrieb daaaasdev <
notifications@github.com>:

**Comment by RehanSaeed** [comment]
This is also affecting System.Text.Json source generated code using `JsonSerializerContext`.

```
[JsonSerializable(typeof(Foo[]))]
public partial class CustomJsonSerializerContext : JsonSerializerContext
{
}
```

**Comment by benmccallum** [comment]
Did this ever get resolved? I've got a situation where I want to ignore this diagnostic under certain namespaces but let it report under others. I tried using a custom diagnostic suppressor but it's the first I've ever written so I'm not sure if I'm doing anything wrong (can it just live in the same project?) or it's just that I'm falling into the same bucket as:
> The global suppressions file can be used to suppress warnings reported by analyzers, but not ones reported by the compiler.

**Comment by mc0re** [comment]
How come the `Add-Migration` tool generates a file starting with `<auto-generated />` marking, but the missing XML documentation still issues an error?

**Comment by letarak** [comment]
Add suppress to the end of editorconfig with desired file mask work for me
```
# resource files
[*.Designer.cs]
dotnet_diagnostic.CS1591.severity = none
```

**Comment by sharwell** [comment]
@DamianSuess Note that we have no control over the dotnet/maui#12196 from this repository.
